src/coreclr/jit/compiler.h

```diff
@@ -8958,6 +8958,20 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
 #endif
     }
 
+    //------------------------------------------------------------------------
+    // canUseEvexEncoding - Answer the question: Is Evex encoding supported on this target.
+    //
+    // Returns:
+    //    TRUE if Evex encoding is supported, FALSE if not.
+    bool canUseEvexEncoding() const
+    {
+#ifdef TARGET_XARCH
+        return compOpportunisticallyDependsOn(InstructionSet_AVX512F);
+#else
+        return false;
+#endif
+    }
```

Comment:
I wonder if we need to differentiate between `EVEX` for `V512` and `EVEX` for `V128/V256` since the two of them are different and the latter requires `AVX512VL`

Reply:
My understanding is **IF** `AVX512F` is available(lowest common denominator to check for and this is true even for `AVX512VL` cases), then we can theoretically use `EVEX` encoding on that target. That's why here(again my assumption here is we are just determining if `EVEX` encoding can be used for this target in this part of code) I check only for `AVX512F`. 

How do you envision differentiating between `EVEX` for V512 and `EVEX` for `V128/V256` at this point in code i.e., is this to figure out if we should use EVEX encoding for individual instruction later  in `TakesEvexPrefix()` or is there another purpose? 

We could have 2 separate checks - 1. `canUseEvex512Encoding` and 2. `canUseEvexVlEncoding` ?. If that's the case, do we add this now or when we add instruction which requires `EVEX` for `V128/V256`? 3. By that logic, shouldn't the same extend for `CD`/`BW`/`DQ`

Reply:
The general consideration here is that unlike VEX where AVX represents "universal encoding support" (that is all legacy encoded instructions now support the VEX encoding), EVEX has a different consideration where it being supported for a given instruction is dependent on AVX512F -or- AV512F+AVX512FVL.

So the concern is that someone may accidentally check `canUseEvexEncoding` and then try to use the encoding to emit the EVEX encoding for a V128/V256 (and therefore VL dependent) instruction.

Particularly with the jit-stress EVEX switch, this may lead to accidental use of VL instructions in an AVX512F only environment.

Reply:
For this @tannergooding , I agree with your thoughts but I am trying to map it to the way `VEX` is handled. Right now, `canUseVexEncoding` only checks for AVX; other parts of the compiler check to use `AVX2`. Can't the same situation happen with `AVX2` right now? Someone checks `canUseVexEncoding` but then emits `AVX2` instructions without actually checking for `AVX2` support?

I agree the jit-stress EVEX switch makes this more of an issue, as it automatically turns on `EVEX` for all supporting instructions (without actually checking support).

Reply:
> Someone checks canUseVexEncoding but then emits AVX2 instructions without actually checking for AVX2 support?

Right, but the difference is that all "prior" instructions support the VEX encoding, so while an AVX2 specific instruction could be emitted accidentally, deciding to "stress or not" is really just about checking AVX.

Where-as for EVEX, all "prior" instructions require AVX512F + AVX512VL.

Maybe limiting EVEX stress to requiring AVX512VL would help limit this concern?

Reply:
Yes, I agree, lets make AVX512F + AVX512VL a requisite for EVEX stress test mode.

@DeepakRajendrakumaran I don't think it is an issue but we should double check to make sure that there are no AVX instructions whose EVEX form depends upon any of the extensions (BW, DQ etc.)

Reply:
So, the consensus seem to be to 'force' EVEX encoding via flag only when both AVX512F + AVX512VL are supported.

We can achieve this in 2 ways
**Option A** : the logic highlighted here is left as it is and the actual change is required in 

```
bool emitter::TakesEvexPrefix(instruction ins) const
{
    if (!JitConfig.JitStressEVEXEncoding() 
               || (JitConfig.JitStressEVEXEncoding() && !compOpportunisticallyDependsOn(InstructionSet_AVX512F_VL))
    {
        return false;
    }
.......
}
```

Note : `compOpportunisticallyDependsOn(InstructionSet_AVX512F_VL)` might need to be propagated in this case

**Option B** : Update `canUseEvexEncoding()`

```
bool canUseEvexEncoding() const
    {
#ifdef TARGET_XARCH
        if (JitConfig.JitStressEVEXEncoding() && !compOpportunisticallyDependsOn(InstructionSet_AVX512F_VL))
                         return false;
        return compOpportunisticallyDependsOn(InstructionSet_AVX512F);
#else
        return false;
#endif
    }
```

In both cases, idea being without the `JitStressEVEXEncoding` flag eventually when we enable this as default, the VL issue is taken care of organically as part of design and that it's only a problem when trying to 'force' EVEX encoding via flag

Reply:
https://github.com/DeepakRajendrakumaran/runtime/blob/avx512-RRR/src/coreclr/jit/compiler.h#L8968-L8981

Have changed to this

Reply:
I think the settled on approach looks great. Thanks!